### DISCUSSION TITLE
Dual Contouring with point's sdf values instead of sdf function

### DISCUSSION THREAD

**Discussion by csyhping** [libigl/libigl#2195] [Q&A] [answered]
Hi @alecjacobson @jdumas , thanks for your amazing work. I'm recently play with igl.dual_contouring, I found that it always require the sdf function f and f_grad as input. What if I just have an array contains sdf values and corresponding normals, how can I use dual contouring in that way? Thanks!

**Comment by alecjacobson** [comment]
What does it mean that you have "corresponding normals" ?

**Reply by csyhping** [reply]
> What does it mean that you have "corresponding normals" ?
hI @alecjacobson , since igl::dual_contouring take [f, f_grad] as input, where f is the sdf function and f_grad is the gradient. What I wanna ask is that, if i have some points with their [sdf_value, gradient], how can i use dual contouring?

**Reply by csyhping** [reply]
In my case, the sdf function is a network model. so i can generate the [sdf_value, gradient] for sampled points, how can I use igl::dual_contouring in such case?

**Reply by alecjacobson** [reply]
can you provide a function handle to evaluate your network and gradient instead of "pre-sampling" points?

**Comment by csyhping** [comment]
Hi @alecjacobson , to make it clearer, what i mean is that: i don't have the sdf function f and f_grad, instead, I have some sampled points and I know the sdf values and gradients on them. How can I use igl::dual contouring in such case? Thanks!

**Reply by alecjacobson** [reply, marked as answer]
right, so igl::dual_contouring needs a function handle to f. If, in general, you just had scattered points, then you'd need to decide on how to interpolate them. It sounds like from your other answer you might be able to make a function around your network to evaluate it as a function.

**Reply by csyhping** [reply]
Yes I agree. I will try to create an interpolation function from these points. Thanks for your answer and help!
